File: ChatNestFullStack/ChatNest/Services/FriendshipService.cs
using ChatNest.Repositories;
using static ChatNest.Models.Domain.Friendship;
using static ChatNest.Services.IFriendshipService;
namespace ChatNest.Services
{
    public class FriendshipService : IFriendshipService
    {
        public FriendshipService(IFriendshipRepository friendshipRepository)
        {
            FriendshipRepository = friendshipRepository;
        }

        public IFriendshipRepository FriendshipRepository { get; }

        public async Task<FriendRequestListResponseModel> GetFriendRequestsAsync(Guid userId, FriendRequestDirection direction)
        {
            var directionString = direction.ToString().ToLower(); // sent or received

            var result = await FriendshipRepository.GetFriendRequestsAsync(userId, directionString);
            return result;
        }

        public async Task<FriendListResponseModel> GetFriendsAsync(Guid userId)
        {
            var response = await FriendshipRepository.GetFriendListAsync(userId);
            return response;
        }

        public async Task<FriendshipStatusResponseModel> GetFriendshipStatusAsync(Guid userId, Guid targetUserId)
        {
            return await FriendshipRepository.CheckFriendshipStatusAsync(userId, targetUserId);
        }

        public async Task<RemoveFriendResponseModel> RemoveFriendAsync(Guid userId, Guid friendId)
        {
            return await FriendshipRepository.RemoveFriendAsync(userId, friendId);
        }

        public async Task<ManageFriendRequestResponseModel> ManageFriendRequestAsync(Guid clientUserID, Guid otherUserID, IFriendshipService.FriendRequestAction action)
        {
            var actionString = action.ToString().ToLower();
            var response = await FriendshipRepository.ManageFriendRequestAsync(clientUserID, otherUserID, actionString);
            return response;
        }

        public async Task<SendFriendRequestResponseModel> SendFriendRequestAsync(Guid requesterId, string receiverEmail)
        {
            var response = await FriendshipRepository.SendFriendRequestAsync(requesterId, receiverEmail);
            return response;
        }
    }
}

File: ChatNestFullStack/ChatNest-app/src/app/features/friendship/services/friendship.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../../api.service';
import { FriendListResponseModel, FriendRequestListResponseModel, FriendshipStatusResponseModel, ManageFriendRequestResponseModel, RemoveFriendResponseModel, SendFriendRequestResponseModel } from '../models/friendship';
import { Observable, tap } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class FriendshipService {

  constructor(
    private ApiService: ApiService
  ) {}

  private baseUrl = '/api/Friendship';

  SendFriendRequestAsync(requesterId:string, recieverEmail:string): Observable<SendFriendRequestResponseModel> {
    return this.ApiService.post<SendFriendRequestResponseModel>(`${this.baseUrl}/SendFriendRequest?requesterId=${requesterId}&recieverEmail=${recieverEmail}`,{}).pipe(
      tap(response => {
        if (response.messageID === 2) {
          console.log('Request re-sent from cancelled state');
        } else if (response.messageID === 3) {
          console.log('Request re-sent from rejected state');
        } else if (response.messageID === 4) {
          console.log('Request re-sent from deleted state');
        }
      })
    );
  }

  GetFriendRequestsAsync(userId:string, direction: 'Sent' | 'Received'): Observable<FriendRequestListResponseModel> {
    return this.ApiService.post<FriendRequestListResponseModel>(`${this.baseUrl}/GetFriendRequests?userId=${userId}&direction=${direction}`, {});    
  }

  GetFriendlistAsync(userId:string): Observable<FriendListResponseModel> {
    return this.ApiService.post<FriendListResponseModel>(`${this.baseUrl}/GetFriendList?userId=${userId}`, {});    
  }

  ManageFriendRequestAsync(clientUserID:string, otherUserID:string, action: 'Accept' | 'Reject' | 'Cancel'): Observable<ManageFriendRequestResponseModel> {
    return this.ApiService.post<ManageFriendRequestResponseModel>(`${this.baseUrl}/ManageFriendRequest?clientUserID=${clientUserID}&otherUserID=${otherUserID}&action=${action}`,{});
  }

  GetFriendshipStatusAsync(userId:string, targetUserId:string): Observable<FriendshipStatusResponseModel> {
    return this.ApiService.post<FriendshipStatusResponseModel>(`${this.baseUrl}/FriendShipStatus?userId=${userId}&targetUserId=${targetUserId}`,{});
  }

  RemoveFriendAsync(userId:string, friendId:string): Observable<RemoveFriendResponseModel> {
    return this.ApiService.post<RemoveFriendResponseModel>(`${this.baseUrl}/RemoveFriend?userId=${userId}&friendId=${friendId}`,{});
  }
}
